File: API/_archive/gRPC/golang/comprehensive_production_grpc.go
package main

import (
	"context"
	"log"
	"net"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
)

// Conceptually assuming we have a generated PB package
// type ServerInterface interface { Unary(ctx), Stream(stream) }

type server struct{}

// 1. Unary RPC (Standard Request/Response with Deadlines & Validation)
func (s *server) ProcessData(ctx context.Context, req interface{}) (interface{}, error) {
	// Validate Data
	// if req.Field == "" { return nil, status.Error(codes.InvalidArgument, "field required") }
	
	// Simulate work respecting context deadline
	select {
	case <-time.After(2 * time.Second):
		return nil, nil // Return response
	case <-ctx.Done():
		log.Println("Context cancelled by client")
		return nil, status.Error(codes.DeadlineExceeded, "deadline exceeded")
	}
}

// 2. Bidirectional Streaming RPC
func (s *server) RealtimeChat(stream interface{}) error { // Concept interface
	// for {
	// 	req, err := stream.Recv()
	// 	if err == io.EOF { return nil }
	// 	if err != nil { return err }
	// 	stream.Send(response)
	// }
	return nil
}

// --- Middleware (Unary Interceptor) ---
func AuthInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
	md, ok := metadata.FromIncomingContext(ctx)
	if !ok || len(md["authorization"]) == 0 {
		return nil, status.Error(codes.Unauthenticated, "missing token")
	}

	token := md["authorization"][0]
	if token != "Bearer secret-jwt" {
		return nil, status.Error(codes.Unauthenticated, "invalid token")
	}

	// Token is valid, proceed to the actual RPC handler
	return handler(ctx, req)
}

func main() {
	lis, err := net.Listen("tcp", ":50051")
	if err != nil {
		log.Fatalf("failed to listen: %v", err)
	}

	// Register Middleware during server startup
	opts := []grpc.ServerOption{
		grpc.UnaryInterceptor(AuthInterceptor),
	}
	
	grpcServer := grpc.NewServer(opts...)
	
	// pb.RegisterMyServiceServer(grpcServer, &server{})
	log.Println("gRPC Production Server started on :50051")
	grpcServer.Serve(lis)
}

File: API/_archive/gRPC/python/comprehensive_production_grpc.py
import asyncio
import grpc

# Conceptually assuming pb2 and pb2_grpc are generated

class MyServiceServicer:
    # 1. Unary RPC
    async def ProcessData(self, request, context: grpc.aio.ServicerContext):
        # Validation
        if not request.field:
            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
            context.set_details("field is required")
            return None
            
        try:
            # Simulate work respecting client deadlines
            await asyncio.wait_for(asyncio.sleep(2), timeout=context.time_remaining())
            return None # Return Response object
        except asyncio.TimeoutError:
            context.abort(grpc.StatusCode.DEADLINE_EXCEEDED, "Time expired")

    # 2. Bidirectional Streaming RPC
    async def RealtimeChat(self, request_iterator, context: grpc.aio.ServicerContext):
        async for request in request_iterator:
            yield None # Yield Response objects

# --- Middleware (Interceptor) ---
class AuthInterceptor(grpc.aio.ServerInterceptor):
    async def intercept_service(self, continuation, handler_call_details):
        metadata = dict(handler_call_details.invocation_metadata)
        
        token = metadata.get("authorization")
        if token != "Bearer secret-jwt":
            return grpc.rpc_method_handlers.unary_unary(
                lambda r, c: c.abort(grpc.StatusCode.UNAUTHENTICATED, "Invalid token")
            )
            
        # Continue to actual handler
        return await continuation(handler_call_details)

async def serve():
    server = grpc.aio.server(interceptors=(AuthInterceptor(),))
    # pb2_grpc.add_MyServiceServicer_to_server(MyServiceServicer(), server)
    
    server.add_insecure_port('[::]:50051')
    print("gRPC Production Server starting on port 50051")
    
    await server.start()
    await server.wait_for_termination()

if __name__ == '__main__':
    # asyncio.run(serve())
    pass
